### PR TITLE
LVI hardening tests

### PR DESCRIPTION
**Pull request by raoulstrackx**
Mitigating the speculative execution LVI attack against SGX enclaves requires compiler changes (i.e., adding lfences). This pull requests adds various tests to check if this happens correctly.

### CODE REVIEW

File: src/test/assembly/x86_64-fortanix-unknown-sgx-lvi-inline-assembly.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+// compile-flags: --crate-type staticlib
+// only-x86_64-fortanix-unknown-sgx
+
+#![feature(llvm_asm)]
+
+#[no_mangle]
+pub extern fn get(ptr: &u64) -> u64 {
+    let value : u64;
+    unsafe {
+        llvm_asm!(".start_inline_asm:
```
Comment (Mark-Simulacrum) [inline src/test/assembly/x86_64-fortanix-unknown-sgx-lvi-inline-assembly.rs:13]: New tests being added should probably be using just the new assembly syntax unless we can't for some reason, could you move this over to asm! instead?

File: ci-rustc.sh [hunk 2]
```diff
@@ -0,0 +1,116 @@
+#!/bin/bash -ex
+
+repo_root=$(readlink -f $(dirname "${BASH_SOURCE[0]}"))
```
Comment (Mark-Simulacrum) [inline ci-rustc.sh:3]: Why was this file added? It looks like it's perhaps something added by mistake?

File: src/bootstrap/native.rs [hunk 3]
```diff
@@ -625,7 +625,15 @@ impl Step for TestHelpers {
         if builder.config.dry_run {
             return;
         }
-        let target = self.target;
+        // The x86_64-fortanix-unknown-sgx target doesn't have a working C
+        // toolchain. However, some x86_64 ELF objects can be linked
+        // without issues. Use this hack to compile the test helpers.
+        let target = if self.target == "x86_64-fortanix-unknown-sgx" {
+            //INTERNER.intern_string("x86_64-unknown-linux-gnu".into()).into()
```
Comment (Mark-Simulacrum) [inline src/bootstrap/native.rs:632]: Please remove the stray commented code here.

**Comment by Mark-Simulacrum** [comment]
It looks like some of this is testing C and C++ LVI, as well, which seems a bit out of place for this repository? Could you elaborate why those tests are added?

File: ci-rustc.sh (hunk 2 above)
Comment (raoulstrackx) [reply ci-rustc.sh:3]: Yes that was a mistake. Removed it

File: src/test/assembly/x86_64-fortanix-unknown-sgx-lvi-inline-assembly.rs (hunk 1 above)
Comment (raoulstrackx) [reply src/test/assembly/x86_64-fortanix-unknown-sgx-lvi-inline-assembly.rs:13]: Moved it to `asm!`

File: src/bootstrap/native.rs (hunk 3 above)
Comment (raoulstrackx) [reply src/bootstrap/native.rs:632]: Fixed

**Comment by raoulstrackx** [comment]
For the `x86_64-fortanix-unknown-sgx` target it is important that _all_ code is hardened that will end up in an SGX enclave. Unfortunately, crates may include assembly, C, C++ code as well that needs to be hardened. We need to check if this is done properly.
Assembly code is certainly compiled as part of the standard library, C/C++ code may not be.

**Comment by Mark-Simulacrum** [comment]
Okay, I think this is reasonable. It's not too many additional tests and they're SGX only so I'm not too worried in that respect either.

@bors r+ rollup=iffy

**Comment by kennytm** [comment]
@bors retry 

https://github.com/rust-lang-ci/rust/runs/1169263815 `i686-msvc-2` failed to download clang.

```
curl: (56) OpenSSL SSL_read: Connection was reset, errno 10054
```
